File: source/downloads/code/LeetCode/Java/Maximum-Product-Subarray.java
public class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        if(n==0)
            return 0;
        int[] pos = new int[n];    
        int[] neg = new int[n];
        pos[0] = nums[0];
        neg[0] = nums[0];

        for(int i=1;i<n;i++){
            pos[i] = Math.max(Math.max(pos[i-1]*nums[i], neg[i-1]*nums[i]), nums[i]);
            neg[i] = Math.min(Math.min(pos[i-1]*nums[i], neg[i-1]*nums[i]), nums[i]);
        }
        int ret = pos[0];
        for(int i=0;i<n;i++){
            if(ret<pos[i])
                ret = pos[i];
        }
        return ret;
    }
}

File: downloads/code/LeetCode/Python/Maximum-Product-Subarray.py
 class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        pos = []  # positive result array
        neg = []  # negative result array

        pos.append(nums[0])
        neg.append(nums[0])
        for i in range(1,len(nums)):
            pos.append(max(pos[i-1]*nums[i], neg[i-1]*nums[i], nums[i]))
            neg.append(min(pos[i-1]*nums[i], neg[i-1]*nums[i], nums[i]))
        # find the largest in pos array
        ret = pos[0]
        for i in pos:
            if ret < i:
                ret = i
        return ret      
